### PR TITLE
Issue 44310: Help guide users to the best authentication option based on their email address

### PR DESCRIPTION
**Pull request by labkey-adam**
#### Rationale
[Issue 44310: Help guide users to the best authentication option based on their email address](https://www.labkey.org/home/Developer/issues/issues-details.view?issueId=44310)

This PR provides user guidance after a failed sign in on the LabKey login page, pointing toward the SSO configuration(s) claiming that user's email domain, For example:

image

#### Related Pull Requests
* https://github.com/LabKey/platform/pull/3018

#### Changes
* Add `LabKeyErrorWithHtml` to convey server-generated HTML to the client
* Login.js: render multiple error messages, if present. Render HTML if provided.
* Add `LabKeyCollectors.joining()` to join streams of `HtmlString`
* Pass attempted email address and return URL to authentication error message generating methods if form-based authentication fails
* Return additional guidance (text and SSO logo(s)) if user's email address is claimed by one or more SSO configurations

### CODE REVIEW

File: api/src/org/labkey/api/security/AuthenticationConfigurationCache.java [hunk 1]
```diff
@@ -138,6 +146,11 @@ private void addToMap(SetValuedMap<Class<? extends AuthenticationConfiguration>,
             return null != configurations ? configurations : Collections.emptyList();
         }
 
+        private @NotNull Collection<AuthenticationConfiguration> getActiveConfigurationsForDomain(String domain)
+        {
+            return new ArrayList<>(_activeDomainMap.get(domain));
```
Comment (labkey-jeckels) [inline api/src/org/labkey/api/security/AuthenticationConfigurationCache.java:151]: Will _activeDomainMap.get() always return non-null here?

**Review by labkey-jeckels — APPROVED** [review]
One question but no need for more review unless you want it

File: api/src/org/labkey/api/security/AuthenticationConfigurationCache.java (hunk 1 above)
Comment (labkey-adam) [reply api/src/org/labkey/api/security/AuthenticationConfigurationCache.java:151]: Yes, that's the contract of `MultiValuedMap.get()`: https://commons.apache.org/proper/commons-collections/apidocs/org/apache/commons/collections4/MultiValuedMap.html#get-K-

Comment (labkey-jeckels) [reply api/src/org/labkey/api/security/AuthenticationConfigurationCache.java:151]: Very good. I missed that it was known to be a MultiValuedMap during the code review

**Pull request merged**